contents/java/JUnit5.md

```diff
@@ -0,0 +1,243 @@
+
+Now, let's fix the code.
+
+```java
+public class Repeater {
+  // Returns copies of reference string separated by the '\n' character.
+  public static String repeat(String reference, int copies) {
+    String result = "";
+    for (int i = 0; i < copies; i++) {
+      result += reference + "\n";
+    }
+    return result;
+  }
+}
+```
+
+<img src="{{baseUrl}}/contents/java/assertPass.PNG" />
+Running the same unit test again, we now have a passing test and we can be <trigger for="modal:somewhatAssured" trigger="click">somewhat assured</trigger> that our repeat function works as intended.
+
+### JUnit APIs
+What else can we test with JUnit? In addition to the `assertEquals` method illustrated above, JUnit allows us to test for many other scenarios. These include
+* `assertTrue`/`assertFalse`: generic checking statements for checking validity.
+* `assertNull`/`assertNotNull`: checks whether something is null or not, without causing a NullPointerException.
+* `assertArrayEquals`: yes, it loops through an entire array for you!
+* `assertThrows`/`assertDoesNotThrow`: checks that a piece of code throws the correct type of exception. In other words, JUnit even checks that the application can handle errors (or not)!
+* `assertTimeout`: ensures that a piece of code runs within time constraints - we don't want to keep the user waiting!
+
+This powerful expression means that we can write more expressive tests.
+Instead of
+```java
+try {
+  int result = divide(100, 0); //Divides 100 by 0
+} catch (ArithmeticException ae) {
+  throw new AssertionError("Divide doesn't work correctly!" + ae);
+}
+```
+We can now write
+```java
+assertThrows(ArithmeticException.class, () -> divide(100, 0));
+```
+
+This expression is a lot easier to understand, and as an added bonus, the code is also much shorter to write!
+
+### Before/After hooks
+
+Not every system can be tested so easily. Sometimes, testing is not as straightforward as running a function.
+For example, we might want to store information in an external database. We would have to first connect to the database before we can run any test.
+
+JUnit 5 exposes an API that allows developers to do this process easily and clearly. It offers the `@BeforeAll` and `@BeforeEach` tags that allow code to be run before each test, allowing us to easily initialise our testing environments before we run any tests:
+
+```java
+class TestInvolvingDatabases {
+  Database db;
+  @BeforeAll
+  public static void initializeDatabaseConnection() {
+    // initialize database connections
+    db = connectToDatabase();
+  }
+  @Test
+  public static void addUserTest() {
+    db.addUser('NewUser');
+    ...
+  }
+}
+```
+
+The `BeforeEach` API can also be leveraged to ensure that we standardise the testing environment before each test.
+For example, we can re-seed a random variable before each test to ensure that we get consistent tests in functions involving randomisation.
```

Comment:
This implicitly assumes that readers would understand what it means to seed a random number generator. Given the target audience of SE students, this may not be a safe assumption to make. Perhaps it would be safer to change this example eg. resetting a in-app database to an initial state?